cosmos/dbt/selector.py

```diff
@@ -84,72 +84,96 @@ def __repr__(self) -> str:
         return f"SelectorConfig(paths={self.paths}, tags={self.tags}, config={self.config}, other={self.other})"
 
 
-def select_nodes_ids_by_intersection(nodes: dict[str, DbtNode], config: SelectorConfig) -> set[str]:
+class NodeSelector:
```

Comment:
The McCabe complexity of this function was 16 because it had a nested function `should_include_node` and several conditional blocks in the nested function. The nested function was needed for caching for the variable `visited_nodes` but inner functions contribute to the complexity. To reduce the complexity I created a class and unnested the `should_include_node` function to it's own method.